### PR TITLE
RFC 0593 Compliance: Allow issuing JSON LD Credentials with credentialStatus

### PR DESCRIPTION
**Pull request by frostyfrog**
Following the documentation for JSONLD Credentials, there is a parameter
for adding `credentialStatus` to a credential:
https://github.com/hyperledger/aries-rfcs/blob/main/features/0593-json-ld-cred-attach/README.md

However, adding the parameter causes all flows to fail due to the
credentialStatus never making it into the credential. I don't know the
history of this parameter, but in order to follow RFC 0593 a bit more
closely, I have fixed up the flow so that it doesn't fail issuance.

Signed-off-by: Colton Wolkins (Indicio work address) <colton@indicio.tech>

### CODE REVIEW

**Comment by swcurran** [comment]
This seems odd. Does this PR enable an implementation of `CredentialStatusList2017` -- including the issuer revoking credentials using that mechanism?   Also -- if we're going to implement support for JSON-LD revocation, we should probably consider using the more recent [Status List 2021](https://w3c-ccg.github.io/vc-status-list-2021/)?

If the full revocation scheme is not implemented, what would a holder receiving a credential with that the CredentialStatus do?

**Comment by dbluhm** [comment]
@swcurran as it stands, the current implementation does not support checking revocation lists directly but it does explicitly prevent their usage through it's current misbehavior. It would seem that it was the intention to support inclusion of the `credentialStatus` in an issued JSON-LD cred but leave the checking of revocation lists up to the controller.

@frostyfrog's changes don't fundamentally change support, they just fix something that we intended to support in the first place, I think.

That's coming from a superficial read of the situation from me, though -- would be interested to hear @TimoGlastra's thoughts! :slightly_smiling_face:

File: aries_cloudagent/messaging/valid.py [hunk 1]
```diff
@@ -757,6 +757,26 @@ def __call__(self, value):
...
+    def __call__(self, value):
+        """Validate input value."""
+        if "type" not in value or value["type"] != CredentialStatus.CREDENTIAL_TYPE:
+            raise ValidationError(
+                f"type must include {CredentialStatus.CREDENTIAL_TYPE}"
+            )
+
+        return value
+
+
```
[the remark this answers is not in the document]
Comment (dbluhm) [reply aries_cloudagent/messaging/valid.py:779]: In other words, let's leave it up to the controller :slightly_smiling_face:

**Comment by swcurran** [comment]
Still a bit leery -- although not sure what to do about the current misbehavior. 

When a credential status is included in a VC, the issuer has to put in both the type of the credential status, and additional information required for the type of status.  For RevocationList2017 (and CredentialStatusList2021), the extra data is (at least) a URI that the verifier can use to check the revocation status using that type. 

I can see including a credential status of "Null" or "None" or what the VC spec. suggests when not using revocation, but I can't see putting in a type, but not a complete implementation.

**Comment by frostyfrog** [comment]
The current implementation accepts anything in the `options` section , but it expects you to have it in the `credential` section during issuance on the holder side. If you put the `credentialStatus` in the `credential` block, then it makes it into the presentation exchange and you fail on the holder side (again). So although the current code supports providing `credentialStatus`, the holder blocks the credential from being received when it's sent from the issuer for one error or another.

File: aries_cloudagent/messaging/valid.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (frostyfrog) [reply aries_cloudagent/messaging/valid.py:779]: That is fair. Do you think that we should check for the existence of "type" and make sure it's a string, with no further validation beyond that?

**Comment by TimoGlastra** [comment]
TBH I'm not sure what my thoughts were at the time, but at the moment I think we should be cautious. 

Having it available but not checked is something the user should know about before using ACA-Py. as with crypto libraries it should be as hard as possible to use it wrong.

Are you currently using credentialStatus for something an implemented the status checks outside of ACA-Py?

Probably the best approach for now would be to at least support verifying it on the verifier side, which is a lot easier than issuance.

Otherwise a property may need to be added whether the revocation was valid/verified -- for indy it will be true, for W3C it will only be true if no credentialStatus was present. Thoughts?

**Comment by dbluhm** [comment]
Thanks everyone that provided feedback! I think it's clear that there's some more work that we want to do before we enable `credentialStatus` usage -- namely, not making security opt-in by default, to put it in just a few words. We (Indicio) are not currently scheduling more work on this problem at the moment. I'm going to close this PR for now and request @frostyfrog to open an issue documenting the desire to add support for `credentialStatus`, current status of that support within ACA-Py, and some of the considerations that have been noted here in the comments.

**Pull request closed without merging**